### ISSUE TITLE
Unexpected behaviour of CrossEntropy Loss function

### ISSUE THREAD

**Issue by jiacheng-xu** [pytorch/pytorch#12670]
## 🐛 Bug

The ignore_index option in the [CrossEntropyLoss](https://pytorch.org/docs/stable/nn.html?highlight=crossen#torch.nn.CrossEntropyLoss) doesn't work as expected.
## To Reproduce

Steps to reproduce the behavior: Run the code:
```
import torch
IGNORE_IDX=  -1
print("PyTorch version: {}".format(torch.__version__))
xe = torch.nn.CrossEntropyLoss(ignore_index=IGNORE_IDX, reduction='none')
input = torch.randn(3, 5, requires_grad=True)
print("Input: {}".format(input))

#### test 1:
target = torch.empty(3, dtype=torch.long).random_(5)
target[2] = -4      # should be an error
print("Target: {}".format(target))
loss = xe(input, target)
print("Loss before reduction: {}".format(loss))
output=torch.mean(loss)
print("Loss after reduction: {}".format(output))
output.backward()
print('-'*20)

#### test 2:
target = torch.empty(3, dtype=torch.long).random_(5)
target[2] = IGNORE_IDX      # should be an expected case
print("Target: {}".format(target))
loss = xe(input, target)
print("Loss before reduction: {}".format(loss))
output=torch.mean(loss)
print("Loss after reduction: {}".format(output))
output.backward()

### test 3: run test 2 again
#### test 2:
target = torch.empty(3, dtype=torch.long).random_(5)
target[2] = IGNORE_IDX      # should be an expected case
print("Target: {}".format(target))
loss = xe(input, target)
print("Loss before reduction: {}".format(loss))
output=torch.mean(loss)
print("Loss after reduction: {}".format(output))
output.backward()
```
Output:
```
PyTorch version: 0.4.1
Input: tensor([[ 0.1701,  0.5413,  0.0602, -0.5931,  0.7988],
        [ 1.3246,  1.7712,  0.3580, -0.7260, -1.5769],
        [-0.7042, -0.6937, -0.3258, -0.2008,  0.5201]], requires_grad=True)
Target: tensor([ 1,  1, -4])
Loss before reduction: tensor([                         1.3669,                          0.6934,
        -4374419384465317223727104.0000], grad_fn=<NllLossBackward>)
Loss after reduction: -1.4581397467833764e+24
--------------------
Target: tensor([ 2,  3, -1])
Loss before reduction: tensor([1.8480, 3.1906, 0.0625], grad_fn=<NllLossBackward>)
Loss after reduction: 1.7003979682922363
Target: tensor([ 0,  4, -1])
Loss before reduction: tensor([                         1.7381,                          4.0416,
        -4409735675994434160295936.0000], grad_fn=<NllLossBackward>)
Loss after reduction: -1.4699119400365407e+24

Process finished with exit code 0

```

## Expected behavior

For the first case: It should throw and error because the target index should range from [0,class]. But it doesn't.
For the second case: The value of the corresponding index should be consistently 0 or (very small number?). But it seems not numerically stable at all. If you have difficulty to reproduce the output, just run the sript for two or three times. 

If i need to manually mask out those invalid bit, the ignore_index option is just confusing and should be removed. I thought the loss i got should just be the average or sum of those valid bits.
## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: TITAN Xp

Nvidia driver version: 384.111
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_static_v6.a
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a
/usr/local/cuda-8.0/lib64/libcudnn.so
/usr/local/cuda-8.0/lib64/libcudnn.so.6
/usr/local/cuda-8.0/lib64/libcudnn.so.6.0.21
/usr/local/cuda-8.0/lib64/libcudnn.so.7
/usr/local/cuda-8.0/lib64/libcudnn.so.7.0.5
/usr/local/cuda-8.0/lib64/libcudnn.so.7.1.1
/usr/local/cuda-8.0/lib64/libcudnn_static.a
/usr/local/cuda-9.0/lib64/libcudnn.so
/usr/local/cuda-9.0/lib64/libcudnn.so.6
/usr/local/cuda-9.0/lib64/libcudnn.so.6.0.21
/usr/local/cuda-9.0/lib64/libcudnn.so.7
/usr/local/cuda-9.0/lib64/libcudnn.so.7.0.5
/usr/local/cuda-9.0/lib64/libcudnn.so.7.1.1
/usr/local/cuda-9.0/lib64/libcudnn_static.a
/usr/local/cuda-9.1/lib64/libcudnn.so
/usr/local/cuda-9.1/lib64/libcudnn.so.6
/usr/local/cuda-9.1/lib64/libcudnn.so.6.0.21
/usr/local/cuda-9.1/lib64/libcudnn.so.7
/usr/local/cuda-9.1/lib64/libcudnn.so.7.0.5
/usr/local/cuda-9.1/lib64/libcudnn.so.7.1.1
/usr/local/cuda-9.1/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] cuda90                    1.0                  h6433d27_0    pytorch
[conda] pytorch                   0.4.1           py36_cuda9.0.176_cudnn7.1.2_1    pytorch
[conda] torch                     0.4.1                     <pip>
[conda] torchfile                 0.1.0                     <pip>
[conda] torchvision               0.2.1                    py36_1    pytorch

```

**Comment by mhubii** [comment]
Why should this line create an error:
```
target = torch.empty(3, dtype=torch.long).random_(5)
target[2] = -4      # should be an error
```
The `target` tensor is of size 3 and you access the last element.

**Comment by weiyangfb** [comment]
not sure if I understand correctly, but sounds like `CrossEntropyLoss` doesn't restrict `class label >= 0`. In your script, `target` gets changed every time, not sure what the code is trying to repro

**Comment by apappu97** [comment]
Hi, was this resolved?

**Comment by mruberry** [comment]
Closing this issue due to age and because ignore_index is documented (it need not be in the class label range, for example).
